### PR TITLE
Add matching module with first draft of MDM

### PR DESCRIPTION
**Pull request by 36000**
Addresses #106 . Adds a first draft of mahalonobis distance matching.
TODO:
- [x] Further modularize this function - break off mahalonobis function, and test it
- [x] raise an error if NaNs in mahalonobis function
- [x] make an additional seaborn like API on top of this

### CODE REVIEW

File: afqinsight/match.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+    Parameters
+    ----------
+    data : pandas Dataframe
+        The feature samples for test subjects.
+    id_column : str
+        Column with unique subject identifiers.
+    test_ids : list of str
+        List of subject identifiers for test subjects.
+    ctrl_ids : list of str
```
Comment (arokem) [inline afqinsight/match.py:26]: I wonder whether there could also be an API where you point to a column (say "is_patient") and then match between groups defined by that column's unique values (has to be a column with only two unique values).

**Review by arokem — COMMENTED** [review]
This is great! Cleanly implemented, tested and documented. 

I have a suggestion for one more  (even higher-level!) API that could be useful to some -- see comment inline. 

And then I think that we should discuss naming: the fact that all the functions have a long name that only differs by the characters after the last underscore suggests that we should make the namespace reflect the commonalities, but I am not 100% sure how to organize this.

**Review by richford — COMMENTED** [review]
Awesome work @36000! I have two smaller suggestions in the review about making the actual mahalonobis distance function private. Seems to me that most users will just want the match.

But I have more substantial comment with a suggestion for the API and want your feedback and @arokem's feedback:

I think we should combine both `mahalonobis_dist_match_df` and `mahalonobis_dist_match` functions and always return a pandas DataFrame.

### Parameters
The expected inputs could be
```
def mahalonobis_dist_match(
    data=None, test=None, ctrl=None,
    id_col=None, status_col="status", feature_columns=None, threshold=0.2
):
```
where if `data is None`, then we assume that `test` and `ctrl` are arrays and use your current `mahalonobis_dist_match` function. But if `data` is provided, then we use `status_col` to match between the groups defined by that columns values as @arokem suggested. If `id_col` is None, we use the index. If `feature_columns is None` then we use all of the columns except `status_col` and `id_col` as the features.

**Comment by richford** [comment]
Looks like there are some linting errors as well. Running `black .` in the repo root should fix that.

**Comment by 36000** [comment]
@arokem @richford I believe I just integrated the changes you both suggested. I like the current API but am also open to further revision. I also removed references to subjectID, as it turned out not to be needed.

File: afqinsight/match.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+            )
+
+        if len(data[status_col].unique()) != 2:
+            raise ValueError(
+                ("status_col in data must " "contain only 2 unique values")
+            )
+
+        if feature_cols is None:
+            feature_cols = data.columns
+            feature_cols.remove(status_col)
```
Comment (richford) [inline afqinsight/match.py:62]: I don't think `remove` is a valid method for `pd.Index` instances. Am I missing something?
```suggestion
            feature_cols = data.columns.drop(status_col)
```

File: afqinsight/match.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
+"""Perform statistical matching comparing test and control units."""
+
+from os import name
```
Comment (richford) [inline afqinsight/match.py:3]: flake suggests getting rid of this
```suggestion
```

File: afqinsight/match.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        If feature_columns is None then we use all of the columns
+        except status_col as the features.
+        Default: None
+    threshold : float, optional
+        Mahalnobis distance beyond which matches are not accepted.
+        Default: None
+
+    Returns
+    -------
+
```
Comment (richford) [inline afqinsight/match.py:47]: flake8 complains that this blank line contains whitespace
```suggestion

```

File: afqinsight/match.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        The feature samples for test subjects.
+        Default: None
+    ctrl : array-like of shape (n_samples2, n_features), optional
+        Feature samples for control subjects from which
+        matches will be selected.
+        Default: None
+    status_col : str, optional
+        Column with subject status. Must have only two unique values,
+        evaluating to True and False, where True indicates a test subject
+        and False indicates a control.
```
Comment (richford) [inline afqinsight/match.py:34]: See suggestion below allowing non boolean values.
```suggestion
        Column with subject status. Must have only two unique values.
```

File: afqinsight/match.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        if len(data[status_col].unique()) != 2:
+            raise ValueError(
+                ("status_col in data must " "contain only 2 unique values")
+            )
+
+        if feature_cols is None:
+            feature_cols = data.columns
+            feature_cols.remove(status_col)
+
+        df_test_rows = data[status_col].astype(bool)
```
Comment (richford) [inline afqinsight/match.py:64]: I suggest letting users can have `status_col` column values like `["test", "control"]` or some such. Then inside the function we can copy the dataframe so as not to change it in place and then label encode the `status_col`:
```suggestion
        data = data.copy()
        data[status_col] = data[status_col].astype("category")
        data[status_col] = data[status_col].cat.codes

        df_test_rows = data[status_col].astype(bool)
```

**Review by richford — COMMENTED** [review]
@36000 Thanks! This is really good stuff. I have a couple more minor comments but I think this is really shaping up and will be very useful.

**Comment by richford** [comment]
Looks like the linting errors have been resolved but now `pydocstyle` is not happy. You can run `pydocstyle` in the root directory to find and fix these locally. Assuming lint and unit tests pass, I think this is good to go and a really great addition to AFQ-Insight. Thanks @36000

**Comment by 36000** [comment]
I think pydocstyle is happy now, thanks for the help!

**Comment by richford** [comment]
@36000, thanks for the substantial development effort on this! Would you mind adding smoke tests for the `ValueError`s raised on current lines 56, 61, and 77 of match.py (e.g. using a `with pytest.raises(ValueError):` context)?

Similarly would you mind adding a unit test for the `feature_cols=None` condition on line 66?

This has been a bit of back and forth with agreeing on the API, linting, and testing and I just want to make sure you know I really appreciate all the work you're putting into this PR. It's a nice enhancement to AFQ-Insight!

**Comment by 36000** [comment]
Thanks for the feedback! I added more testing.

**Comment by richford** [comment]
I think it's `black --check .` that's failing. BTW, if you do `pre-commit install` then it will install the pre-commit hook that does all the lint checking before committing.

**Comment by 36000** [comment]
I missed that part of the contributing file, my bad, let's see if this works...
